tests/ed25519.rs

```diff
@@ -116,81 +130,154 @@ mod vectors {
         );
         assert!(
             signing_key
-                .verify_prehashed(prehash_for_verifying, None, &sig2)
+                .verify_prehashed(prehash_for_verifying.clone(), None, &sig2)
                 .is_ok(),
             "Could not verify ed25519ph signature!"
         );
+        assert!(
+            expected_verifying_key
+                .verify_prehashed_strict(prehash_for_verifying, None, &sig2)
+                .is_ok(),
+            "Could not strict-verify ed25519ph signature!"
+        );
     }
 
+    //
+    // The remaining items in this mod are for the repudiation tests
+    //
+
     // Taken from curve25519_dalek::constants::EIGHT_TORSION[4]
     const EIGHT_TORSION_4: [u8; 32] = [
         236, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
         255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 127,
     ];
 
-    fn compute_hram(message: &[u8], pub_key: &EdwardsPoint, signature_r: &EdwardsPoint) -> Scalar {
-        let k_bytes = Sha512::default()
-            .chain_update(&signature_r.compress().as_bytes())
-            .chain_update(&pub_key.compress().as_bytes()[..])
-            .chain_update(&message);
-        let mut k_output = [0u8; 64];
-        k_output.copy_from_slice(k_bytes.finalize().as_slice());
-        Scalar::from_bytes_mod_order_wide(&k_output)
+    // Computes the prehashed or non-prehashed challenge, depending on whether context is given
+    fn compute_challenge(
+        message: &[u8],
+        pub_key: &EdwardsPoint,
+        signature_r: &EdwardsPoint,
+        context: Option<&[u8]>,
+    ) -> Scalar {
+        let mut h = Sha512::default();
+        if let Some(c) = context {
+            h.update(b"SigEd25519 no Ed25519 collisions");
+            h.update(&[1]);
+            h.update(&[c.len() as u8]);
+            h.update(c);
+        }
+        h.update(&signature_r.compress().as_bytes());
+        h.update(&pub_key.compress().as_bytes()[..]);
+        h.update(&message);
+        Scalar::from_hash(h)
     }
 
     fn serialize_signature(r: &EdwardsPoint, s: &Scalar) -> Vec<u8> {
         [&r.compress().as_bytes()[..], &s.as_bytes()[..]].concat()
     }
 
+    const WEAK_PUBKEY: CompressedEdwardsY = CompressedEdwardsY(EIGHT_TORSION_4);
+
+    // Pick a random Scalar
+    fn non_null_scalar() -> Scalar {
+        let mut rng = rand::rngs::OsRng;
+        let mut s_candidate = Scalar::random(&mut rng);
+        while s_candidate == Scalar::ZERO {
+            s_candidate = Scalar::random(&mut rng);
+        }
+        s_candidate
+    }
+
+    fn pick_r(s: Scalar) -> EdwardsPoint {
+        let r0 = s * ED25519_BASEPOINT_POINT;
```

Comment:
After #251 lands we should probably look for places like this where `mul_base` could be used